libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Trace.hs

```diff
@@ -0,0 +1,194 @@
+    Show (State (EraRule "PPUP" era))
+  )
+
+type ToCBORCore era = (ToCBOR (Core.TxBody era), ToCBOR (Core.TxOut era), ToCBOR (Core.Value era))
+
+type LEDGERRule era =
+  ( Environment (EraRule "DELEGS" era) ~ DelegsEnv era,
+    Signal (EraRule "UTXOW" era) ~ Core.Tx era,
+    State (EraRule "UTXOW" era) ~ UTxOState era,
+    Environment (EraRule "UTXOW" era) ~ UtxoEnv era,
+    State (EraRule "DELEGS" era) ~ DPState (Crypto era),
+    Signal (EraRule "DELEGS" era) ~ Seq (DCert (Crypto era)),
+    Embed (EraRule "DELEGS" era) (LEDGER era),
+    Embed (EraRule "UTXOW" era) (LEDGER era),
+    Embed (EraRule "UTXOW" era) (LEDGER era),
+    HasField "_keyDeposit" (Core.PParams era) Coin,
+    HasField "_poolDeposit" (Core.PParams era) Coin,
+    HasField "certs" (Core.TxBody era) (StrictSeq (DCert (Crypto era))),
+    HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era)))
+  )
+
+-- ==============================================================
+-- HasTrace instance of LEDGER
+
+instance
+  ( Reflect era,
+    GoodCrypto (Crypto era),
+    LEDGERRule era,
+    ToCBORCore era,
+    ShowCore era
+  ) =>
+  HasTrace (LEDGER era) (GenState era)
+  where
+  type BaseEnv (LEDGER era) = Globals
+
+  interpretSTS globals act = runIdentity $ runReaderT act globals
+
+  envGen gstate =
+    LedgerEnv (getSlot gstate)
+      <$> arbitrary
+      <*> pure ((gePParams . gsGenEnv) gstate)
+      <*> pure
+        ( AccountState
+            (getTreasury gstate)
+            (getReserves gstate)
+        )
+
+  sigGen genstate _ledgerenv (utxostate, _dpstate) = genTx genstate utxostate
+
+  shrinkSignal _x = []
+
+-- | We make this a toplevel function because we will have to call it
+--   in the LEDGERS Rule as well.
+genTx :: Reflect era => GenState era -> UTxOState era -> Gen (Core.Tx era)
+genTx genstate utxostate =
+  runReaderT
+    (snd <$> (genValidatedTx (gsProof genstate)))
+    (genstate {gsPrevUTxO = _utxo utxostate})
+
+-- ====================================
+-- HasTrace instance of LEDGERS
+
+type LEDGERSRule era =
+  ( Signal (EraRule "LEDGER" era) ~ Core.Tx era,
+    State (EraRule "LEDGER" era) ~ (UTxOState era, DPState (Crypto era)),
+    Environment (EraRule "LEDGER" era) ~ LedgerEnv era,
+    Embed (EraRule "LEDGER" era) (LEDGERS era),
+    Default (State (EraRule "PPUP" era))
+  )
+
+instance
+  ( Reflect era,
+    GoodCrypto (Crypto era),
+    LEDGERSRule era,
+    Signal (EraRule "LEDGER" era) ~ Core.Tx era
+  ) =>
+  HasTrace (LEDGERS era) (GenState era)
+  where
+  type BaseEnv (LEDGERS era) = Globals
+
+  interpretSTS globals act = runIdentity $ runReaderT act globals
+
+  envGen gstate =
+    pure $
+      LedgersEnv
+        (getSlot gstate)
+        ((gePParams . gsGenEnv) gstate)
+        ( AccountState
+            (getTreasury gstate)
+            (getReserves gstate)
+        )
+
+  -- a LEDGERS signal is a sequence of LEDGER signals (Seq (ValidatedTx era))
+  sigGen genstate ledgersenv (LedgerState u dp) = do
```

Comment:
For the sake of Haskell conventions maybe renaming them to: `genState` and `ledgersEnv`? 
Just a suggestion, feel free to ignore it.

Reply:
this code has been deleted.